gui-agent/vmside.c

```diff
@@ -1505,54 +1509,64 @@ static void mkghandles(Ghandles * g)
-    g->wmProtocols = XInternAtom(g->display, "WM_PROTOCOLS", False);
-    g->wm_hints = XInternAtom(g->display, "WM_HINTS", False);
-    g->wm_class = XInternAtom(g->display, "WM_CLASS", False);
-    g->utf8_string_atom = XInternAtom(g->display, "UTF8_STRING", False);
     g->stub_win = XCreateSimpleWindow(g->display, g->root_win,
             0, 0, 1, 1,
             0, BlackPixel(g->display,
                 g->screen),
             WhitePixel(g->display,
                 g->screen));
+    if ((unsigned)snprintf(tray_sel_atom_name, sizeof(tray_sel_atom_name),
+            "_NET_SYSTEM_TRAY_S%u", DefaultScreen(g->display)) >=
+        sizeof tray_sel_atom_name)
+        abort();
     /* pretend that GUI agent is window manager */
-    g->net_wm_name = XInternAtom(g->display, "_NET_WM_NAME", False);
-    net_supporting_wm_check = XInternAtom(g->display, "_NET_SUPPORTING_WM_CHECK", False);
-    net_supported = XInternAtom(g->display, "_NET_SUPPORTED", False);
-    supported[0] = net_supported;
-    supported[1] = net_supporting_wm_check;
-    /* _NET_WM_MOVERESIZE required to disable broken GTK+ move/resize fallback */
-    supported[2] = XInternAtom(g->display, "_NET_WM_MOVERESIZE", False);
-    supported[3] = XInternAtom(g->display, "_NET_WM_STATE", False);
-    supported[4] = XInternAtom(g->display, "_NET_WM_STATE_FULLSCREEN", False);
-    supported[5] = XInternAtom(g->display, "_NET_WM_STATE_DEMANDS_ATTENTION", False);
-    XChangeProperty(g->display, g->stub_win, g->net_wm_name, g->utf8_string_atom,
-            8, PropModeReplace, (unsigned char*)"Qubes", 5);
-    XChangeProperty(g->display, g->stub_win, net_supporting_wm_check, XA_WINDOW,
-            32, PropModeReplace, (unsigned char*)&g->stub_win, 1);
-    XChangeProperty(g->display, g->root_win, net_supporting_wm_check, XA_WINDOW,
-            32, PropModeReplace, (unsigned char*)&g->stub_win, 1);
-    XChangeProperty(g->display, g->root_win, net_supported, XA_ATOM,
-            32, PropModeReplace, (unsigned char*)supported, sizeof(supported)/sizeof(supported[0]));
+    const char * atom_names[] = {
+        "_NET_SUPPORTED",
+        "_NET_SUPPORTING_WM_CHECK",
+        /* _NET_WM_MOVERESIZE required to disable broken GTK+ move/resize fallback */
+        "_NET_WM_MOVERESIZE",
+        "_NET_WM_STATE",
+        "_NET_WM_STATE_FULLSCREEN",
+        "_NET_WM_STATE_DEMANDS_ATTENTION",
+#define X(a, b) b,
+        GLOBAL_ATOMS(X, tray_sel_atom_name)
+#undef X
+    };
+    Atom supported[QUBES_ARRAY_SIZE(atom_names)];
+    if (!XInternAtoms(g->display, (char **)atom_names, QUBES_ARRAY_SIZE(atom_names),
+                      False, supported)) {
+        fputs("Could not intern global atoms\n", stderr);
+        exit(1);
+    }
+#define FIRST(a) FIRST2(a)
+#define FIRST2(a, ...) a
+#define X(a, b) a,
+    memcpy((unsigned char *)g + offsetof(Ghandles, FIRST(GLOBAL_ATOMS(X,))),
+           supported + 6,
+           sizeof supported - 6 * sizeof supported[0]);
```

Comment:
This is too much macro magic, plus a memcpy for filling a structure fields gives very little room for compiler to spot any mistakes.
I'd prefer the way you did it in gui-daemon.

Reply:
Will do.  Using XCB instead of Xlib might also be a good idea.